experimental/dds/tree2/src/feature-libraries/modular-schema/typedSchema/flexList.ts

```diff
@@ -0,0 +1,94 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+// TODO: tests
+
+/**
+ * Flexible way to list values.
+ * Can be values, functions that return the value (to allow cyclic references to work), or arrays.
+ * @remarks
+ * Does not work properly if T can be a function or array.
+ * @alpha
+ */
+export type FlexList<Item = unknown> = readonly LazyItem<Item>[];
+
+export function normalizeFlexList<List extends FlexList>(t: List): FlexListToLazyArray<List> {
+	return t.map((value: LazyItem) => {
+		if (typeof value === "function") {
+			return value;
+		}
+		return () => value;
+	}) as FlexListToLazyArray<List>;
+}
+
+export function normalizeFlexListEager<List extends FlexList>(
+	t: List,
+): FlexListToNonLazyArray<List> {
+	return t.map((value: LazyItem) => {
+		if (typeof value === "function") {
+			return value() as unknown;
+		}
+		return value;
+	}) as FlexListToNonLazyArray<List>;
+}
+
+/**
+ * T, but can be wrapped in a function to allow referring to types before they are declared.
+ * This makes recursive and co-recursive types possible.
+ * @alpha
+ */
+export type LazyItem<Item = unknown> = Item | (() => Item);
+
+/**
+ * @alpha
+ */
+export type NormalizedFlexList<Item> = readonly Item[];
+
+export type NormalizedLazyFlexList<Item> = (() => Item)[];
+
+/**
+ * @alpha
+ */
+export type ExtractItemType<List extends LazyItem> = List extends () => infer Result
+	? Result
+	: List;
+type NormalizeLazyItem<List extends LazyItem> = List extends () => unknown ? List : () => List;
+
+/**
+ * Normalize FlexList type to a non-lazy array.
+ * @alpha
+ */
+export type FlexListToNonLazyArray<List extends FlexList> = number extends List["length"]
+	? // Handle non compile time constant case
+	  NormalizedFlexList<List extends FlexList<infer Item> ? Item : unknown>
+	: ConstantFlexListToNonLazyArray<List>;
```

Comment:
THats right. I've refactoerd it to be a bit more clear and add some tests since it is strange.

making stronly typed tuples readonly mainly adds noise to the type names. You already can't assign their members to anything that violates the very specific types or change the length.

Also its annoying since variadic argument lists (which are how we capture allowed types) alwayse end up not readonly, and having some of them readonly and some not depending on if they have been nromalized messes up one of the tests.